lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,284 @@
+import           Formatting (oct, sformat)
+import qualified System.Posix.Files as PSX
+import qualified System.Posix.Types as PSX (FileMode)
+import           System.Wlog (logWarning)
+#endif
+
+-- Because of the Formatting import
+{-# ANN module ("HLint: ignore Use fewer imports" :: Text) #-}
+
+-- | Describes HD wallets keyfile content
+data WalletUserPublic = WalletUserPublic
+    { _wupWalletName :: Text                -- ^ name of wallet
+    , _wupAccounts   :: [(Word32, Text)]    -- ^ accounts coordinates and names
+    , _wupAddrs      :: [(Word32, Word32)]  -- ^ addresses coordinates
+    } deriving (Eq, Generic, Show)
+
+makeLenses ''WalletUserPublic
+
+instance Buildable WalletUserPublic where
+    build WalletUserPublic{..} =
+        bprint ("{ wallet name = "%build%
+                ", accounts = "%pairsF%", addresses = "%pairsF%" }")
+        _wupWalletName
+        _wupAccounts
+        _wupAddrs
+      where
+        pairsF :: (Buildable a, Buildable b) => Format r ([(a, b)] -> r)
+        pairsF = later $ mconcat . map (uncurry $ bprint ("("%build%", "%build%")"))
+
+deriveSimpleBi ''WalletUserPublic [
+    Cons 'WalletUserPublic [
+        Field [| _wupWalletName :: Text               |],
+        Field [| _wupAccounts   :: [(Word32, Text)]   |],
+        Field [| _wupAddrs      :: [(Word32, Word32)] |]
+    ]]
+
+-- | User public data.
+data UserPublic = UserPublic
+    { _upKeys   :: [PublicKey]
+    , _upWallet :: Maybe WalletUserPublic
+    , _upPath   :: FilePath
+    , _upLock   :: Maybe FileLock
+    } deriving (Eq, Generic)
+
+makeLenses ''UserPublic
+
+class HasUserPublic ctx where
+    -- if you're going to mock this TVar, look how it's done for peer state.
+    userPublic :: Lens' ctx (TVar UserPublic)
+
+-- | Show instance to be able to include it into NodeParams
+instance Bi Address => Show UserPublic where
+    show UserPublic {..} =
+        formatToString
+            ("UserPublic { _upKeys = "%listJson%
+             ", _upPath = "%build%", _upWallet = "%build%"}")
+            _upKeys
+            _upPath
+            _upWallet
+
+newtype UserPublicDecodingError = UserPublicDecodingError Text
+    deriving (Show)
+
+instance Exception UserPublicDecodingError
+
+instance Buildable UserPublicDecodingError where
+    build (UserPublicDecodingError msg) =
+        "Failed to decode user public: " <> bprint build msg
+
+-- | Path of lock file for the provided path.
+lockFilePath :: FilePath -> FilePath
+lockFilePath = (<> ".lock")
+
+-- | Checks if this user public instance can be dumped back to
+-- file. If not, using 'writeUserPublic' and 'writeUserPublicRelease'
+-- will result in error.
+canWrite :: UserPublic -> Bool
+canWrite up = up ^. upLock . to isJust
+
+instance Default UserPublic where
+    def = UserPublic [] Nothing "" Nothing
+
+-- | It's not network/system-related, so instance shouldn't be under
+-- @Pos.Binary.*@.
+instance Bi UserPublic where
+  encode us = encodeListLen 2 <> encode (_upKeys us)
```

Comment:
Firstly, I do realise that this data type is not used on the network or on the blockchain.

However, the Core Team has spent considerable effort writing golden and  round trip tests for other `Bi` instances to make sure they don't get inadvertently broken when the code is modified. 

Since this seems to be on-disk storage of key information, I wonder if it might not be a good idea to add version information to the binary representation so that if the format ever needs to be changed, the old format and the new format can be easily distinguished.